0055-rrna/rrna.py

```diff
@@ -0,0 +1,120 @@
+
+parser.add_argument(
+    '--outFile', required=True, metavar='FILENAME',
+    help='The filename the output will be written to.')
+
+args = parser.parse_args()
+
+mappedReads = FastqReads(args.mappedFile)
+unmappedReads = FastqReads(args.unmappedFile)
+coverageDepthFile = args.coverageDepthFile
+outfile = args.outFile
+
+# The Lengths below in bp, taken from NR_146117 genbank annotation.
+# sExonLength45 = 13502
+# sRrnaLength18 = 1872
+# sRrnaLength5 = 156
+# sRrnaLength28 = 5195
+# sDnaLength45 = sExonLength45 - sRrnaLength18 - sRrnaLength5 - sRrnaLength28
+
+# These are the positions, 0-based, taken from NR_146117 genbank annotation.
+sRrna18 = (3659, 5531)
+sRrna5 = (6615, 6771)
+sRrna28 = (7942, 13137)
+sRna45 = (sRrna18, sRrna5, sRrna28)
+sDna45 = ((0, 3659), (5531, 6615), (6771, 7942), (13137, 13502))
+
+
+def countReads(reads):
+    """
+    Given a reads instance, count the number of reads.
+
+    @param reads: A darkmatter reads instance.
+    @return: A C{float} of the number of reads.
+    """
+    count = 0
+    for read in reads:
+        count += 1
+
+    return count
+
+
+def calcPercent(partial, total):
+    """
+    Given two numbers, give the percentage.
+
+    @param partial: Number of to divide.
+    @param total: Number of.total.
+    @return: A C{float} giving the %.
+    """
+    return partial / total * 100
+
+
+def averageCoverageDepth(coverageDepthFile, region):
+    """
+    Given an outputfile from sam-coverage-depth.py and a list of regions, give
+    the average coverage-depth over those regions.
+
+    @param coverageDepthFile: sam-coverage-depth outfile.
+    @param region: a C{tuples} with a range of interest.
+    @return: A C{float} giving the average coverage depth across the range.
+    """
+    with open(coverageDepthFile) as cd:
+        depthList = []
+        for line in cd:
+            for index in range(region[0], region[1]):
+                if str(index) == line.split(':')[0]:
+                    depthList.append(int(line.split()[1]))
+        if len(depthList) == 0:
+            averageCoverage = 0
+        else:
+            averageCoverage = sum(depthList) / len(depthList)
+        return averageCoverage
+
+
+nbMapped = countReads(mappedReads)
+nbAll = nbMapped + countReads(unmappedReads)
+
+with open(outfile, 'w') as fp:
+    mappedPercent = calcPercent(nbMapped, nbAll)
+
+    allRrna = [averageCoverageDepth(coverageDepthFile, rRNA)
+               for rRNA in sRna45]
+    averageRrna = sum(allRrna) / len(allRrna) if allRrna else 0.0
+
+    allRdna = [averageCoverageDepth(coverageDepthFile, rDNA)
+               for rDNA in sDna45]
+    averageRdna = sum(allRdna) / len(allRdna) if allRdna else 0.0
+
+    if averageRrna == 0:
+        print('No reads mapped against rRNA! Exiting...', file=sys.stderr)
+        exit()
+
+    correctionCoeff = averageRdna / averageRrna
+
+    print('Percent of mapped reads, corrected for rDNA mapping:', file=fp)
```

Comment:
The `with open(outfile, 'w') as fp:` line could come right before these two lines. The way you have the code atm, you force the reader (who will be you at some future point) to keep in mind that we're in the context of a file opened for writing 18 lines of code ago. So as I read the code I keep wondering when you're going to use `fp`, and finally you do. Better to put all this together. These tiny things really do add some cognitive overload to reading and reasoning about the code.  It matters little in isolated instances but it's good practice to keep the logic tighter (i.e., not so separated on the page, in the code) and easier to keep track of for more complicated situations.

Reply:
yes, makes sense. Done